services/src/main/java/io/grpc/services/HealthServiceImpl.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.services;
+
+import static io.grpc.health.v1.HealthCheckResponse.ServingStatus;
+
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.health.v1.HealthCheckRequest;
+import io.grpc.health.v1.HealthCheckResponse;
+import io.grpc.health.v1.HealthGrpc;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.Nullable;
+
+
+final class HealthServiceImpl extends HealthGrpc.AbstractHealth {
+
+  /* Due to the latency of rpc calls, synchronization of the map does not help with consistency.
+   * However, need use ConcurrentHashMap to prevent the possible race condition of currently putting
+   * two keys with a colliding hashCode into the map. */
+  private final Map<String, ServingStatus> statusMap
+      = new ConcurrentHashMap<String, ServingStatus>();
+
+  @Override
+  public void check(HealthCheckRequest request,
+      StreamObserver<HealthCheckResponse> responseObserver) {
+    ServingStatus status = getStatus(request.getService());
+    if (status == null) {
+      responseObserver.onError(new StatusException(Status.NOT_FOUND));
```

Comment:
@dapengzhang0 i am a bit confused about the default behavior here

from the grpc health check doc:
> A client can query the server’s health status by calling the Check method, and a deadline should be set on the rpc. The client can **OPTIONALLY** set the service name it wants to query for health status.

with the current implementation, the server returns a Status.NOT_FOUND for an empty service check. shouldn't it respond with the general health of the server, which should be counted as SERVING independent of the included service's statuses.

Reply:
nvm: the very next statement from the doc helped clarify it:

> The server should register all the services manually and set the individual status, including an empty service name and its status.